--- ../src-base/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
@@ -1,3 +1,5 @@
+// rpm0618 (Reverted carpet changes entirely)
+
 package net.minecraft.world.chunk.storage;
 
 import com.google.common.collect.Maps;
@@ -37,11 +39,12 @@
 public class AnvilChunkLoader implements IChunkLoader, IThreadedFileIO
 {
     private static final Logger field_151505_a = LogManager.getLogger();
-    private final Map<ChunkPos, NBTTagCompound> field_75828_a = Maps.<ChunkPos, NBTTagCompound>newConcurrentMap();
-    private final Set<ChunkPos> field_193415_c = Collections.<ChunkPos>newSetFromMap(Maps.newConcurrentMap());
-    private final File field_75825_d;
+    private final Map<ChunkPos, NBTTagCompound> chunksToRemove = Maps.<ChunkPos, NBTTagCompound>newConcurrentMap();
+    private final Set<ChunkPos> field_193415_c = Collections.newSetFromMap(Maps.newConcurrentMap());
+
+    public final File field_75825_d;
     private final DataFixer field_193416_e;
-    private boolean field_183014_e;
+    private boolean savingExtraData;
 
     public AnvilChunkLoader(File p_i46673_1_, DataFixer p_i46673_2_)
     {
@@ -50,10 +53,11 @@
     }
 
     @Nullable
+
     public Chunk func_75815_a(World p_75815_1_, int p_75815_2_, int p_75815_3_) throws IOException
     {
         ChunkPos chunkpos = new ChunkPos(p_75815_2_, p_75815_3_);
-        NBTTagCompound nbttagcompound = this.field_75828_a.get(chunkpos);
+        NBTTagCompound nbttagcompound = (NBTTagCompound)this.chunksToRemove.get(chunkpos);
 
         if (nbttagcompound == null)
         {
@@ -73,11 +77,12 @@
     public boolean func_191063_a(int p_191063_1_, int p_191063_2_)
     {
         ChunkPos chunkpos = new ChunkPos(p_191063_1_, p_191063_2_);
-        NBTTagCompound nbttagcompound = this.field_75828_a.get(chunkpos);
+        NBTTagCompound nbttagcompound = (NBTTagCompound)this.chunksToRemove.get(chunkpos);
         return nbttagcompound != null ? true : RegionFileCache.func_191064_f(this.field_75825_d, p_191063_1_, p_191063_2_);
     }
 
     @Nullable
+
     protected Chunk func_75822_a(World p_75822_1_, int p_75822_2_, int p_75822_3_, NBTTagCompound p_75822_4_)
     {
         if (!p_75822_4_.func_150297_b("Level", 10))
@@ -120,13 +125,13 @@
             NBTTagCompound nbttagcompound = new NBTTagCompound();
             NBTTagCompound nbttagcompound1 = new NBTTagCompound();
             nbttagcompound.func_74782_a("Level", nbttagcompound1);
-            nbttagcompound.func_74768_a("DataVersion", 1343);
+            nbttagcompound.func_74768_a("DataVersion", 1139);
             this.func_75820_a(p_75816_2_, p_75816_1_, nbttagcompound1);
             this.func_75824_a(p_75816_2_.func_76632_l(), nbttagcompound);
         }
         catch (Exception exception)
         {
-            field_151505_a.error("Failed to save chunk", (Throwable)exception);
+            field_151505_a.error("Failed to save chunk", exception);
         }
     }
 
@@ -134,7 +139,7 @@
     {
         if (!this.field_193415_c.contains(p_75824_1_))
         {
-            this.field_75828_a.put(p_75824_1_, p_75824_2_);
+            this.chunksToRemove.put(p_75824_1_, p_75824_2_);
         }
 
         ThreadedFileIOBase.func_178779_a().func_75735_a(this);
@@ -142,24 +147,33 @@
 
     public boolean func_75814_c()
     {
-        if (this.field_75828_a.isEmpty())
+        if (this.chunksToRemove.isEmpty())
         {
-            if (this.field_183014_e)
+            if (this.savingExtraData)
             {
-                field_151505_a.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", (Object)this.field_75825_d.getName());
+                field_151505_a.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", this.field_75825_d.getName());
             }
 
             return false;
         }
         else
         {
-            ChunkPos chunkpos = this.field_75828_a.keySet().iterator().next();
+            ChunkPos chunkpos = (ChunkPos)this.chunksToRemove.keySet().iterator().next();
             boolean lvt_3_1_;
 
+            // rpm0618
+            if (this.chunksToRemove.size() % 1000 == 0) {
+                System.out.println("Save queue size: " + this.chunksToRemove.size());
+            }
+            if (this.chunksToRemove.size() == 1) {
+                System.out.println("Save queue: Saving final chunk");
+            }
+            // end rpm0618
+
             try
             {
                 this.field_193415_c.add(chunkpos);
-                NBTTagCompound nbttagcompound = this.field_75828_a.remove(chunkpos);
+                NBTTagCompound nbttagcompound = (NBTTagCompound)this.chunksToRemove.remove(chunkpos);
 
                 if (nbttagcompound != null)
                 {
@@ -169,7 +183,7 @@
                     }
                     catch (Exception exception)
                     {
-                        field_151505_a.error("Failed to save chunk", (Throwable)exception);
+                        field_151505_a.error("Failed to save chunk", exception);
                     }
                 }
 
@@ -178,6 +192,15 @@
             finally
             {
                 this.field_193415_c.remove(chunkpos);
+
+                // rpm0618
+                if (chunkpos.field_77276_a == 352 && chunkpos.field_77275_b == -370) {
+                    System.out.println("Chunk Saving Done: " + chunkpos);
+                }
+                if (chunkpos.field_77276_a == 352 && chunkpos.field_77275_b == -369) {
+                    System.out.println("Chunk Saving Done: " + chunkpos);
+                }
+                // end rpm0618
             }
 
             return lvt_3_1_;
@@ -203,13 +226,13 @@
     {
         try
         {
-            this.field_183014_e = true;
+            this.savingExtraData = true;
 
             while (this.func_75814_c());
         }
         finally
         {
-            this.field_183014_e = false;
+            this.savingExtraData = false;
         }
     }
 
@@ -249,7 +272,7 @@
         });
     }
 
-    private void func_75820_a(Chunk p_75820_1_, World p_75820_2_, NBTTagCompound p_75820_3_)
+    public static void func_75820_a(Chunk p_75820_1_, World p_75820_2_, NBTTagCompound p_75820_3_)
     {
         p_75820_3_.func_74768_a("xPos", p_75820_1_.field_76635_g);
         p_75820_3_.func_74768_a("zPos", p_75820_1_.field_76647_h);
